src/network/src/esp/src/NetworkManager.cpp

```diff
@@ -0,0 +1,142 @@
+    ESP_ERROR_CHECK(esp_wifi_init(&cfg));
+
+    // Initialize default station as network interface instance. Will change in the future when
+    // enabling mesh capabilities
+    auto* networkInterface = esp_netif_create_default_wifi_sta();
+    assert(networkInterface);
+}
+
+void NetworkManager::eventHandler(void* context,
+                                  esp_event_base_t eventBase,
+                                  int32_t eventId,
+                                  void* eventData) {
+    (void)context;
+    auto* manager = (NetworkManager*)(&NetworkContainer::getNetworkManager());
+
+    if (eventBase == WIFI_EVENT && eventId == WIFI_EVENT_STA_START) {
+        manager->m_logger.log(LogLevel::Error, "Started wifi, attempting to connect...");
+        esp_wifi_connect();
+    } else if (eventBase == WIFI_EVENT && eventId == WIFI_EVENT_STA_DISCONNECTED) {
+        manager->m_logger.log(LogLevel::Error, "Network error, attempting to reconnect");
+        manager->m_state = NetworkManagerState::DISCONNECTED;
+    } else if (eventBase == IP_EVENT && eventId == IP_EVENT_STA_GOT_IP) {
+        auto* event = (ip_event_got_ip_t*)eventData;
+        manager->m_state = NetworkManagerState::CONNECTED;
+        manager->m_logger.log(LogLevel::Info, "Network manager got ip:" IPSTR,
+                              IP2STR(&event->ip_info.ip));
+        manager->m_ipAddress.u_addr.ip4 = event->ip_info.ip;
+    }
+}
+
+void NetworkManager::start() {
+    initNetworkInterface();
+    m_networkExecuteTask.start();
+}
+
+NetworkStatus NetworkManager::getNetworkStatus() {
+    switch (m_state) {
+    case NetworkManagerState::LOOKING_FOR_NETWORK:
+        return NetworkStatus::Connecting;
+    case NetworkManagerState::CONNECTED:
+    case NetworkManagerState::RUNNING:
+        return NetworkStatus::Connected;
+
+    case NetworkManagerState::INIT:
+    case NetworkManagerState::DISCONNECTED:
+    default:
+        return NetworkStatus::NotConnected;
+    }
+}
+
+void NetworkManager::getNetworkingID(std::string& id) {
+    char buff[16];
+    esp_ip4_addr_t ip = getIP();
+    sprintf(buff, IPSTR, IP2STR(&ip));
+    id.assign(buff);
+}
+
+void NetworkManager::execute() {
+    char message[] = "Test message";
+    switch (m_state) {
+
+    case NetworkManagerState::INIT:
+        m_logger.log(LogLevel::Info, "Connecting to network...");
+        ESP_ERROR_CHECK(esp_wifi_set_mode(NetworkConfig::getMode()));
+        ESP_ERROR_CHECK(esp_wifi_set_config(NetworkConfig::getInterface(),
+                                            NetworkConfig::getDefaultNetworkConfig()));
+        ESP_ERROR_CHECK(esp_wifi_start());
+
+        m_state = NetworkManagerState::LOOKING_FOR_NETWORK;
+        break;
+    case NetworkManagerState::LOOKING_FOR_NETWORK:
+        // Idle state
+        break;
+    case NetworkManagerState::CONNECTED:
+        m_logger.log(LogLevel::Info, "Connected to network!");
+
+        if (m_server.start()) {
```

Comment:
The goal of the network manager is to start the network_interface and then connect to the network based on a configuration. Lwip sockets cannot be attempted to be created beforehand, otherwise, the device will crash and reboot. This behavior was designed by ESP.  The network manager provides a way of knowing the status of the network. If the device is not connected, it is useless to try to get a socket. 
Upon connecting to a network, the network manager then signals the server that it can start. The start function of the server obtains a socket. Then, if the connection is lost, the server socket is closed, until it can be restarted again when rejoining the network.